src/vm/dllimport.cpp

```diff
@@ -6132,22 +6136,107 @@ bool         NDirect::s_fSecureLoadLibrarySupported = false;
 #endif // !FEATURE_PAL
 
 // static
-NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryFromPath(LPCWSTR libraryPath)
+NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryFromPath(LPCWSTR libraryPath, BOOL throwOnError)
 {
     STANDARD_VM_CONTRACT;
 
     LoadLibErrorTracker errorTracker;
     const NATIVE_LIBRARY_HANDLE hmod =
         LocalLoadLibraryHelper(libraryPath, GetLoadWithAlteredSearchPathFlag(), &errorTracker);
-    if (hmod == nullptr)
+    
+    if (throwOnError && (hmod == nullptr))
     {
         SString libraryPathSString(libraryPath);
         errorTracker.Throw(libraryPathSString);
     }
     return hmod;
 }
 
-/* static */
+// static 
+NATIVE_LIBRARY_HANDLE NDirect::LoadLibraryByName(LPCWSTR libraryName, Assembly *callingAssembly, 
+                                                 DWORD dllImportSearchPathFlag, BOOL throwOnError)
+{
+    STANDARD_VM_CONTRACT;
+
+    LoadLibErrorTracker errorTracker;
+
+    BOOL searchAssemblyDirectory = dllImportSearchPathFlag & DLLIMPORTSEARCHPATH_ASSEMBLYDIRECTORY;
+    dllImportSearchPathFlag &= ~DLLIMPORTSEARCHPATH_ASSEMBLYDIRECTORY;
+
+    const NATIVE_LIBRARY_HANDLE hmod = 
+        LoadLibraryModuleBySearch(callingAssembly, searchAssemblyDirectory, dllImportSearchPathFlag, &errorTracker, libraryName);
+
+    if (throwOnError && (hmod == nullptr))
+    {
+        SString libraryPathSString(libraryName);
+        errorTracker.Throw(libraryPathSString);
+    }
+
+    return hmod;
+}
+
+// static
+void NDirect::FreeNativeLibrary(NATIVE_LIBRARY_HANDLE handle)
+{
+    STANDARD_VM_CONTRACT;
+
+    if (handle == NULL)
+        COMPlusThrowArgumentNull(W("handle"), W("Arg_InvalidHandle"));
+
+#ifndef FEATURE_PAL
+    BOOL retVal = FreeLibrary(handle);
+#else // !FEATURE_PAL
+    BOOL retVal = PAL_FreeLibraryDirect(handle);
+#endif // !FEATURE_PAL
+
+    if (retVal == 0)
+        COMPlusThrow(kInvalidOperationException, W("Arg_InvalidOperationException"));
+}
+
+//static 
+INT_PTR NDirect::GetNativeLibraryExport(NATIVE_LIBRARY_HANDLE handle, LPCWSTR symbolName, BOOL throwOnError)
+{
+    STANDARD_VM_CONTRACT;
+
+    INT_PTR address = NULL;
+
+    if(handle == NULL) 
+    {
+        if(throwOnError)
+            COMPlusThrowArgumentNull(W("handle"), W("Arg_InvalidHandle"));
+        return NULL;
+    }
+
+    if(symbolName == NULL)
+    {
+        if(throwOnError)
+            COMPlusThrowArgumentNull(W("symbolName"), W("ArgumentNull_String"));
+        return NULL;
+    }
+
+    MAKE_UTF8PTR_FROMWIDE_NOTHROW(lpstr, symbolName);
+
+    if(lpstr == NULL)
+    {
+        if(throwOnError)
```

Comment:
The convention for `TryXXX` APIs is to throw for invalid input. They should skip throwing exception only if the operation itself fails.

For example, `Dictionary.TryGetValue(null, out ...)` is throws `ArgumentNullException`.